src/bucket/Bucket.cpp

```diff
@@ -146,69 +128,220 @@ Bucket::convertToBucketEntry(std::vector<LedgerKey> const& deadEntries)
+            "unsupported entry type {} in protocol {} bucket",
+            (entry.type() == INITENTRY ? "INIT" : "META"), protocolVersion));
     }
-    return bucket;
 }
 
 inline void
 maybePut(BucketOutputIterator& out, BucketEntry const& entry,
-         std::vector<BucketInputIterator>& shadowIterators)
+         std::vector<BucketInputIterator>& shadowIterators,
+         bool keepShadowedLifecycleEntries, MergeCounters& mc)
 {
     BucketEntryIdCmp cmp;
     for (auto& si : shadowIterators)
     {
         // Advance the shadowIterator while it's less than the candidate
         while (si && cmp(*si, entry))
         {
+            ++mc.mShadowScanSteps;
             ++si;
         }
         // We have stepped si forward to the point that either si is exhausted,
         // or else *si >= entry; we now check the opposite direction to see if
         // we have equality.
         if (si && !cmp(entry, *si))
         {
-            // If so, then entry is shadowed in at least one level and we will
-            // not be doing a 'put'; we return early. There is no need to
-            // advance the other iterators, they will advance as and if
+            // If so, then entry is shadowed in at least one level and we
+            // will potentially not be doing a 'put', but rather returning
+            // early. Whether or not we do a 'put' here, there is no need
+            // to advance the other iterators, they will advance as and if
             // necessary in future calls to maybePut.
-            return;
+            //
+            // In ledgers before protocol 11, keepShadowedLifecycleEntries
+            // will be `false` and we will drop all shadowed entries here.
+            //
+            // In ledgers at-or-after protocol 11, it will be `true` which
+            // means that we only elide 'put'ing an entry if it is in
+            // LIVEENTRY state; we keep entries in DEADENTRY and INITENTRY
+            // states, for two reasons:
+            //
+            //   - DEADENTRY is preserved to ensure that old live-or-init
+            //     entries that were killed remain dead, are not brought
+            //     back to life accidentally by having a newer shadow
+            //     eliding their later DEADENTRY (tombstone). This is
+            //     possible because newer shadowing entries may both refer
+            //     to the same key as an older dead entry, and may occur as
+            //     an INIT/DEAD pair that subsequently annihilate one
+            //     another.
+            //
+            //     IOW we want to prevent the following scenario:
+            //
+            //       lev1:DEAD, lev2:INIT, lev3:DEAD, lev4:INIT
+            //
+            //     from turning into the following by shadowing:
+            //
+            //       lev1:DEAD, lev2:INIT, -elided-, lev4:INIT
+            //
+            //     and then the following by pairwise annihilation:
```

Comment:
a comment somewhere would be welcome to explain the following.

took me a while to figure this one out: the reason you need to use the max version including shadows is to avoid annihilating a DEAD entry when merging two pre-protocol 11 buckets.

you would have the same problem than the one written here even but instead of `lev4:INIT` you would have a `lev4:LIVE` (that would never get deleted)

Reply:
Good point! I actually already forgot that was a strict requirement and put in a wimpier comment. Comment on versions-derived-from-shadows augmented.